envoy/network/address.h

```diff
@@ -230,6 +231,39 @@ class Instance {
    * @return SocketInterface to be used with the address.
    */
   virtual const Network::SocketInterface& socketInterface() const PURE;
+
+  /**
+   * For support the absl::Hash.
+   */
+  virtual void abslHashValue(absl::HashState state) const PURE;
+};
+
+/**
+ * Wrapper for InstanceConstSharedPtr, enable to use InstanceConstSharedPtr
+ * hash key for absl::Hash.
+ */
+struct AddressKey {
+  AddressKey(InstanceConstSharedPtr address) : address_(address) {}
+
+  friend bool operator==(const AddressKey& lhs, const AddressKey& rhs) {
+    // For the envoy internal address, only compare the address id and
+    // ignore the endpoint id.
+    if (lhs.address_->type() == Type::EnvoyInternal &&
```

Comment:
I think it would be simpler to also hash endpoint_id. Semantically, this is similar to a "port" on an address, and it's reasonable to expect that:
* upstream specifies an internal "port" to an internal address.
* downstream listens on all "ports" for an internal address.

The same code path should work for non-internal address if we used actual ports, and it's better to avoid exceptions.

Since we don't support individual "endpoint_id" on downstream, we can add a check that denies internal downstream listen address with a non-empty endpoint_id.

Reply:
> I think it would be simpler to also hash endpoint_id. Semantically, this is similar to a "port" on an address, and it's reasonable to expect that:
> 
> * upstream specifies an internal "port" to an internal address.
> * downstream listens on all "ports" for an internal address.
> 
> The same code path should work for non-internal address if we used actual ports, and it's better to avoid exceptions.

got it, +1 for hash all type addresses similar 

> 
> Since we don't support individual "endpoint_id" on downstream, we can add a check that denies internal downstream listen address with a non-empty endpoint_id.

got it, I can submit a separated PR for this, I guess we need a runtime flag for this, right?

Reply:
@kyessenov gentle ping

Reply:
I think it's fine to strip endpoint_id when searching for a matching listener if we have to have one big registry. Alternatively, we'd just keep the map separate for internal addresses. I don't recall which way is more natural at this point.

Reply:
got it, Iet me try to strip endpoint id, thanks